src/basic/process-util.c

```diff
@@ -2215,3 +2215,26 @@ _noreturn_ void report_errno_and_exit(int errno_fd, int error) {
 
         _exit(EXIT_FAILURE);
 }
+
+int read_errno(int errno_fd) {
+        int r;
+
+        assert(errno_fd >= 0);
+
+        ssize_t n = loop_read(errno_fd, &r, sizeof(r), /* do_pool = */ 0);
+        if (n < 0) {
+                log_debug_errno(n, "Failed to read errno: %m");
+                return -EIO;
+        }
+        if (n == sizeof(r)) {
+                /* child processes reported a error, return it */
+                if (r < 0)
+                        return log_debug_errno(r, "Child process failed with errno: %m");
+                return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Received a errno, but it's a positive value");
+        }
+        if (n != 0)
+                return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Received unexpected amount of bytes while reading errno");
+
+        /* the process exited without reporting a error, assuming success */
+        return 0;
```

Comment:
so if a child dies abnormally it would close such an errno fd, and we might mistake that as success. maybe alter the protocol a bit: insist we send a 0 on success, and if we don't we assume there's something wrong and propagate EPROTO or so.

Reply:
I thought about it. What stopped me from doing so is that there is a bunch of code where `report_errno_and_exit()` is used but it does no reporting on success. For instance: https://github.com/systemd/systemd/blob/main/src/core/exec-invoke.c#L2257

I imagined `read_errno()` to be used together with `report_errno_and_exit()`. So, if we follow your suggestion there would be a need to change logic in the described cases.

Reply:
ok, so actually we don't need to signal success, we have the exit status for that of the child process, so maybe my concerns make no sense.